addon/services/prefetch.js

```diff
@@ -25,14 +25,16 @@ if (gte('3.6.0')) {
             let changeSet = createPrefetchChangeSet(privateRouter, transition);
             if (changeSet.shouldCall) {
               for (let i = 0; i < changeSet.for.length; i++) {
-                let { route, fullParams } = changeSet.for[i];
-                if (seenRoutes.has(route)) continue;
-
-                  route._prefetched = new RSVP.Promise(r => {
-                    return r(route.prefetch(fullParams, transition))
+                // normalize route to always be a promise since lazy loaded
+                // routes are promises, while preloaded routes are not. We want
+                // to treat them both the same.
+                RSVP.resolve(changeSet.for[i]).then(({ route, fullParams }) => {
```

Comment:
As far as I can tell, `changeSet.for` can never be a promise.

`createPrefetchChangeSet` does not use promises, and returns a synchronous value:

https://github.com/nickiaconis/ember-prefetch/blob/fc5a60281536129e57e240a0b6eb15c565382b3f/addon/-private/diff-route-info.js#L127-L157

The `changeSet.for` value is the result of calling `getPrefetched` here, which does not use promises either:

https://github.com/nickiaconis/ember-prefetch/blob/fc5a60281536129e57e240a0b6eb15c565382b3f/addon/-private/diff-route-info.js#L114-L125

Reply:
yes, you are 100% right about that and it does look odd that I would do this, but I had some intuition that without this we could end up using an unexpected `fullParams` object inside the promise `then` handler since it is pulls this out of changeSet.for synchronously in the loop before the callback runs. I could be mistaken here, but it seemed to me like there could be a situation where we end up using the `fullParams` from the last loop iteration when we don't really mean to. There may be other ways to address this as now we are iterating the entire changeSet anyway and we don't ever break early like the old implementation was doing. Or maybe my intuition was just wrong and I'm not thinking clearly about it.

Reply:
I actually changed the implementation a bit to handle promise case and synchronous case separately. Let me know what you think.

Reply:
I liked having a generic way of handling things, but I think the trade off there is that its not really clear why certain things are being done. I think the new approach makes it much more explicit what is happening and why.

Reply:
made some other minor changes. I think we can use the promise object directly in seenRoutes WeekMap. I don't think it really matters what type of object is used there.